### PR TITLE
add dot-product attention

### PR DESCRIPTION
**Pull request by ranqiu92**
@lcy-seso

### CODE REVIEW

File: python/paddle/trainer_config_helpers/networks.py [hunk 1]
```diff
@@ -1396,6 +1398,85 @@ def simple_attention(encoded_sequence,
...
+
+    The example usage is:
+
+    ..  code-block:: python
+
+        context = dot_product_attention(encoded_sequence=enc_seq,
+                                        attending_sequence=att_seq,
+                                        transformed_state=state,)
+
+    :param name: name of the dot-product attention model.
```
Comment (lcy-seso) [inline python/paddle/trainer_config_helpers/networks.py]: A prefix attached to the name of each layer that defined inside the dot_product_attention.

File: python/paddle/trainer_config_helpers/networks.py [hunk 2]
```diff
@@ -1396,6 +1398,85 @@ def simple_attention(encoded_sequence,
...
+    :math:`s_{i-1}` is transformed_state
+
+    The example usage is:
+
+    ..  code-block:: python
+
+        context = dot_product_attention(encoded_sequence=enc_seq,
+                                        attending_sequence=att_seq,
+                                        transformed_state=state,)
+
```
Comment (lcy-seso) [inline python/paddle/trainer_config_helpers/networks.py]: In all the comments below, please make each individual sentence begins with the capital letter and ends with a period.

File: python/paddle/trainer_config_helpers/networks.py [hunk 3]
```diff
@@ -1396,6 +1398,85 @@ def simple_attention(encoded_sequence,
...
+    :param name: name of the dot-product attention model.
+    :type name: basestring
+    :param softmax_param_attr: parameter attribute of sequence softmax
```
Comment (lcy-seso) [inline python/paddle/trainer_config_helpers/networks.py]: The parameter attribute ....

File: python/paddle/trainer_config_helpers/networks.py [hunk 4]
```diff
@@ -1396,6 +1398,85 @@ def simple_attention(encoded_sequence,
...
+                               that is used to produce attention weight.
+    :type softmax_param_attr: ParameterAttribute
+    :param encoded_sequence: output of the encoder
```
Comment (lcy-seso) [inline python/paddle/trainer_config_helpers/networks.py]: The output hidden vectors of the encoder.

File: python/paddle/trainer_config_helpers/networks.py [hunk 5]
```diff
@@ -1396,6 +1398,85 @@ def simple_attention(encoded_sequence,
...
+    :type encoded_sequence: LayerOutput
+    :param attending_sequence: attention weight is computed by a feed forward neural
```
Comment (lcy-seso) [inline python/paddle/trainer_config_helpers/networks.py]: The attention weight ...

File: python/paddle/trainer_config_helpers/networks.py [hunk 6]
```diff
@@ -1396,6 +1398,85 @@ def simple_attention(encoded_sequence,
...
+                               network which has two inputs : decoder's transformed
+                               hidden state of previous time step and encoder's output.
+                               attending_sequence is the sequence to be attended.
+    :type attending_sequence: LayerOutput
+    :param transformed_state: transformed hidden state of decoder in previous time step,
```
Comment (lcy-seso) [inline python/paddle/trainer_config_helpers/networks.py]: - The transformed ...
- Are words "transformed hidden state" the commonly accepted name used in the original paper?

File: python/paddle/trainer_config_helpers/networks.py [hunk 7]
```diff
@@ -1396,6 +1398,85 @@ def simple_attention(encoded_sequence,
...
+                              its size should equal to encoded_sequence's. Here we do the
```
Comment (lcy-seso) [inline python/paddle/trainer_config_helpers/networks.py]: whose dimension should be equal to encoded_sequence's dimension.  Or use a period at the end of last sentence and changes "its" into "Its".

File: python/paddle/trainer_config_helpers/networks.py [hunk 8]
```diff
@@ -1396,6 +1398,85 @@ def simple_attention(encoded_sequence,
...
+                              transformation outside dot_product_attention for flexibility
+                              consideration.
+    :type transformed_state: LayerOutput
+    :return: a context vector
```
Comment (lcy-seso) [inline python/paddle/trainer_config_helpers/networks.py]: The context vector.

File: python/paddle/trainer_config_helpers/networks.py [hunk 9]
```diff
@@ -1396,6 +1398,85 @@ def simple_attention(encoded_sequence,
...
+    :rtype: LayerOutput
+    """
+    assert transformed_state.size == encoded_sequence.size
```
Comment (lcy-seso) [inline python/paddle/trainer_config_helpers/networks.py]: please leaves a message to explain the check.

File: python/paddle/trainer_config_helpers/networks.py [hunk 10]
```diff
@@ -1396,6 +1398,85 @@ def simple_attention(encoded_sequence,
         input=scaled, pooling_type=SumPooling(), name="%s_pooling" % name)
 
 
+@wrap_name_default()
+def dot_product_attention(encoded_sequence,
+                          attending_sequence,
```
[the remark this answers is not in the document]
Comment (ranqiu92) [reply python/paddle/trainer_config_helpers/networks.py]: Done

File: python/paddle/trainer_config_helpers/networks.py [hunk 11]
```diff
@@ -1396,6 +1398,85 @@ def simple_attention(encoded_sequence,
...
+                          transformed_state,
+                          softmax_param_attr=None,
+                          name=None):
+    """
+    Calculate and return a context vector with dot-product attention mechanism.
+    Size of the context vector equals to size of the attending_sequence.
```
[the remark this answers is not in the document]
Comment (ranqiu92) [reply python/paddle/trainer_config_helpers/networks.py]: Done

File: python/paddle/trainer_config_helpers/networks.py [hunk 12]
```diff
@@ -1396,6 +1398,85 @@ def simple_attention(encoded_sequence,
...
+        a(s_{i-1},h_{j}) & = s_{i-1}^\mathrm{T} h_{j}
+
+        e_{i,j} & = a(s_{i-1}, h_{j})
+
+        a_{i,j} & = \\frac{exp(e_{i,j})}{\\sum_{k=1}^{T_x}{exp(e_{i,k})}}
+
+        c_{i} & = \\sum_{j=1}^{T_{x}}a_{i,j}z_{j}
+
+    where :math:`h_{j}` is the jth element of encoded_sequence,
+    :math:`z_{j}` is the jth element of attending_sequence,
```
[the remark this answers is not in the document]
Comment (ranqiu92) [reply python/paddle/trainer_config_helpers/networks.py]: Done

File: python/paddle/trainer_config_helpers/networks.py [hunk 13]
```diff
@@ -1396,6 +1398,85 @@ def simple_attention(encoded_sequence,
...
+    :math:`s_{i-1}` is transformed_state
+
+    The example usage is:
+
+    ..  code-block:: python
+
+        context = dot_product_attention(encoded_sequence=enc_seq,
+                                        attending_sequence=att_seq,
```
[the remark this answers is not in the document]
Comment (ranqiu92) [reply python/paddle/trainer_config_helpers/networks.py]: Done

File: python/paddle/trainer_config_helpers/networks.py (hunk 2 above)
Comment (ranqiu92) [reply python/paddle/trainer_config_helpers/networks.py]: Done

File: python/paddle/trainer_config_helpers/networks.py (hunk 1 above)
Comment (ranqiu92) [reply python/paddle/trainer_config_helpers/networks.py]: Done

File: python/paddle/trainer_config_helpers/networks.py (hunk 3 above)
Comment (ranqiu92) [reply python/paddle/trainer_config_helpers/networks.py]: Done

File: python/paddle/trainer_config_helpers/networks.py (hunk 4 above)
Comment (ranqiu92) [reply python/paddle/trainer_config_helpers/networks.py]: Done

File: python/paddle/trainer_config_helpers/networks.py (hunk 5 above)
Comment (ranqiu92) [reply python/paddle/trainer_config_helpers/networks.py]: Done

File: python/paddle/trainer_config_helpers/networks.py (hunk 8 above)
Comment (ranqiu92) [reply python/paddle/trainer_config_helpers/networks.py]: Done

File: python/paddle/trainer_config_helpers/networks.py (hunk 9 above)
Comment (ranqiu92) [reply python/paddle/trainer_config_helpers/networks.py]: Done

File: python/paddle/trainer_config_helpers/networks.py (hunk 7 above)
Comment (ranqiu92) [reply python/paddle/trainer_config_helpers/networks.py]: Done

File: python/paddle/trainer_config_helpers/networks.py (hunk 6 above)
Comment (ranqiu92) [reply python/paddle/trainer_config_helpers/networks.py]: i use this just for flexibility consideration

**Pull request merged**